### PR TITLE
at86rf2xx: fix receive before send detection

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

The at86rf2xx radio handled a transfer complete condition with the radio in the BUSY_TX_ARET state as a finished transmission. This condition and state also occurs when a reception occurs just before switching to transmitting. This would cause a condition where first a TX_COMPLETE was
signalled and second a RX_COMPLETE was signalled. The network stack would then read the transmitted frame as a received frame.

The patch fixes the errornous RX callback by only submitting the TX_COMPLETE condition when there are at least 2 frames pending (at86rf2xx::pending_tx).

This should not impact packet loss with the radio. The frame received is also discarded without this patch.

The change set is not that big, it just appears big due to splitting out the tx complete handling from the `_isr()` call. **Edit**: splitted the tx complete refactor to a separate commit to ease reviewing.

### Testing procedure

The test as supplied with #11256 must pass. `examples/gnrc_networking` should work as usual on a samr21-xpro or on the iotlab-m3,

### Issues/PRs references

Fixes the issue described in #11256

### CODE REVIEW

**Comment by jia200x** [comment]
This is also fixed in #12069 , but I'm fine to go with this approach since #12069 will make more sense with the netif rework. I will give it a deeper look

**Comment by bergzand** [comment]
> I will give it a deeper look

Thanks!

**Comment by bergzand** [comment]
ping @jia200x

**Comment by benpicco** [comment]
Murdock/clang are not happy.

File: drivers/at86rf2xx/at86rf2xx_netdev.c [hunk 1]
```diff
@@ -667,54 +712,28 @@ static void _isr(netdev_t *netdev)
...
+            at86rf2xx_set_state(dev, dev->idle_state);
+            DEBUG("[at86rf2xx] return to idle state 0x%x\n", dev->idle_state);
+            _isr_send_complete(dev, trac_status);
+        }
+        /* Only the case when an interrupt was received and the radio is busy
+         * with a next PDU transmission when _isr is called.
+         * dev->pending == 1 means a receive and immediately a send happened.
+         * The receive is discarded as the send already overwrote the internal
+         * buffer.
+         * dev->pending == 2 means two transmits occured and this is the isr for
```
Comment (benpicco) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]:
```suggestion
         * dev->pending == 2 means two transmits occurred and this is the isr for
```

Comment (bergzand) [reply drivers/at86rf2xx/at86rf2xx_netdev.c]: Fixed!

File: drivers/at86rf2xx/at86rf2xx_netdev.c [hunk 2]
```diff
@@ -667,54 +712,28 @@ static void _isr(netdev_t *netdev)
             }
             netdev->event_callback(netdev, NETDEV_EVENT_RX_COMPLETE);
         }
-        else if ((state == AT86RF2XX_STATE_TX_ARET_ON)
-                 || (state == AT86RF2XX_STATE_BUSY_TX_ARET)) {
+        else if ((state == AT86RF2XX_STATE_TX_ARET_ON)) {
```
Comment (benpicco) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]:
```suggestion
        else if (state == AT86RF2XX_STATE_TX_ARET_ON) {
```

File: drivers/at86rf2xx/at86rf2xx_netdev.c [hunk 3]
```diff
@@ -667,54 +712,28 @@ static void _isr(netdev_t *netdev)
...
+        }
+        /* Only the case when an interrupt was received and the radio is busy
+         * with a next PDU transmission when _isr is called.
+         * dev->pending == 1 means a receive and immediately a send happened.
+         * The receive is discarded as the send already overwrote the internal
+         * buffer.
+         * dev->pending == 2 means two transmits occurred and this is the isr for
+         * the first.
+         */
+        else if ((state == AT86RF2XX_STATE_BUSY_TX_ARET)) {
```
Comment (benpicco) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]:
```suggestion
        else if (state == AT86RF2XX_STATE_BUSY_TX_ARET) {
```

**Comment by bergzand** [comment]
Thanks @benpicco!

**Pull request merged**